### PR TITLE
Generate virtual player

### PR DESCRIPTION
**Pull request by kcwojdzinska**

### CODE REVIEW

File: scripts/generate_virtual_player.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+import os
+
+from player import VirtualPlayer
+from game_session import GameSession
+from models import RescorlaWagner, Qlearning
+
+def prepare_parameters(number_of_trials, num_of_params):
+    parameters = np.zeros((number_of_trials, num_of_params))
+    if num_of_params == 3:
+        for params in range(parameters.shape[0]):
```
Comment (jczestochowska) [inline scripts/generate_virtual_player.py]: notice than you have a separate variable for expression parameters.shape[0],
if statement is redundant as you use the same distribution both for case when 2 parameters are needed and for the case with 3, you could just use slicing e.g parameters[params][1:] = random.uniform(). Notice also that naming index of array "params" is somewhat misleading for the reader.

File: scripts/generate_virtual_player.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+    return parameters
+
+
+def _create_file_path(filename):
+    parent_directory = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+    return os.path.join(parent_directory, "data", filename)
+
+
+if __name__=='__main__':
```
Comment (jczestochowska) [inline scripts/generate_virtual_player.py]: tihs script lacks function that would write as many csv files as there are rows in 'parameters' to seperate directory

File: scripts/generate_virtual_player.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        for params in range(parameters.shape[0]):
+            parameters[params][0] = random.random()
+            parameters[params][1] = random.uniform(-10,10)
+
+    return parameters
+
+
+def _create_file_path(filename):
+    parent_directory = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+    return os.path.join(parent_directory, "data", filename)
```
Comment (jczestochowska) [inline scripts/generate_virtual_player.py]: better to include "data" as parameter, then this function could be moved to data utils and used 'projectwide'

File: scripts/generate_virtual_player.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    params = prepare_parameters(10,3)
+    attempt = params[0,:]
+    player = VirtualPlayer(game.game_skeleton, attempt[0], attempt[1], attempt[2])
+    model = RescorlaWagner()
+    game.play(player=player, model=model)
+    game._create_result()
+    file = np.array2string(attempt)
+    path = _create_file_path(filename="RW"+file)
+    game.result.to_csv(path, index=False)
+    #print(params[0,:][0])
```
Comment (jczestochowska) [inline scripts/generate_virtual_player.py]: please avoid commented lines as they could be merged accidentally

File: scripts/generate_virtual_player.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    game = GameSession()
+    params = prepare_parameters(10,3)
+    attempt = params[0,:]
+    player = VirtualPlayer(game.game_skeleton, attempt[0], attempt[1], attempt[2])
+    model = RescorlaWagner()
+    game.play(player=player, model=model)
+    game._create_result()
+    file = np.array2string(attempt)
+    path = _create_file_path(filename="RW"+file)
+    game.result.to_csv(path, index=False)
```
Comment (jczestochowska) [inline scripts/generate_virtual_player.py]: you can simply use save_results method from GameSession, then invoking private method _create_result wouldnt be necessary after all it is private for a reason :p

File: scripts/generate_virtual_player.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+def _create_file_path(filename):
+    parent_directory = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+    return os.path.join(parent_directory, "data", filename)
+
+
+if __name__=='__main__':
+    game = GameSession()
+    params = prepare_parameters(10,3)
+    attempt = params[0,:]
```
Comment (jczestochowska) [inline scripts/generate_virtual_player.py]: attempt is also a misleading name here, why would you create 10 sets of parameters and then use only one? this regards also comment about lacking function for generating multiple csv

File: scripts/generate_virtual_player.py (hunk 3 above)
Comment (jczestochowska) [reply scripts/generate_virtual_player.py]: as for this, method in GameSession named _create_file_path should be rewritten as well

**Review by jczestochowska — DISMISSED** [review]

File: scripts/generate_game_sessions.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+def prepare_params():
+    T = np.arange(-10, 10, 0.5).tolist()
+    alpha = np.arange(-1, 1, 0.1).tolist()
+    params_ql = itertools.product(T, alpha)
+    params_rl = itertools.product(T, alpha, alpha)
+    return params_ql, params_rl
+
+
+def generate_game_sessions_with_real_player_parameters():
+    make_directory(directory='/Users/karola/PycharmProjects/ZPI/data/virtual_players')
```
Comment (jczestochowska) [inline scripts/generate_game_sessions.py]: hardcoding paths is not a good idea

File: scripts/generate_game_sessions.py [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    df = pd.read_csv('/Users/karola/PycharmProjects/ZPI/scripts/new_real_player_params_ql.csv')
+    for index, row in df.iterrows():
+        game = GameSession()
+        model = Qlearning()
+        # model = RescorlaWagner()
```
Comment (jczestochowska) [inline scripts/generate_game_sessions.py]: including commented lines not a good idea

File: scripts/generate_game_sessions.py [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+        game._create_result()
+        file = tuple(map(str, (round_number(row['T']), round_number(row['alpha']), row['name'])))
+        # file = tuple(map(str, (round_number(row['T']), round_number(row['alpha_gain'], round_number(row['alpha_lose']), row['name'])))
+        path = _create_file_path(filename="QL_" + "_".join(file))
+        # path = _create_file_path(filename="RW_"+"_".join(file))
+        game.result.to_csv(path, index=False)
+
+
+def generate_game_sessions_with_all_parameters():
+    make_directory(directory='/Users/karola/PycharmProjects/ZPI/data/virtual_players')
```
Comment (jczestochowska) [inline scripts/generate_game_sessions.py]: same here

**Review by jczestochowska — CHANGES_REQUESTED** [review]

**Review by dudamarlena — APPROVED** [review]

**Review by dudamarlena — APPROVED** [review]

**Review by jczestochowska — APPROVED** [review]

**Pull request merged**